src/pybind/mgr/ansible/module.py

```diff
@@ -178,19 +195,96 @@ def is_effective(self):
         """
 
-        return self.status == ExecutionStatusCode.SUCCESS
+        return self._status == ExecutionStatusCode.SUCCESS
 
     @property
     def is_errored(self):
-        return self.error
+        return self._status == ExecutionStatusCode.ERROR
 
     @property
-    def is_complete(self):
-        return self.is_errored or (self.is_persistent and self.is_effective)
+    def result(self):
+        return self._result
+
+class HttpOperation(object):
+
+    def __init__(self, url, http_operation, payload="", query_string="{}"):
+        """ A class to ease the management of http operations
+        """
+        self.url = url
+        self.http_operation = http_operation
+        self.payload = payload
+        self.query_string = query_string
+        self.response = None
+
+class ARSChangeOperation(AnsibleChangeOperation):
+    """Execute one or more Ansible Runner Service Operations that implies
+    a change in the cluster
+    """
+    def __init__(self, client, logger, operations):
+        """
+        :param client         : Ansible Runner Service Client
+        :param logger         : The object used to log messages
+        :param operations     : A list of http_operation objects
+        :param payload        : dict with http request payload
+        """
+        super(ARSChangeOperation, self).__init__()
+
+        assert operations, "At least one operation is needed"
+        self.ar_client = client
+        self.log = logger
+        self.operations = operations
+
+        self.process_output = None
+
+    def __str__(self):
+             # Use the last operation as the main
+             return "Ansible Runner Service: {operation} {url}".format(
+                operation = self.operations[-1].http_operation,
+                url = self.operations[-1].url)
+
+    @property
+    def status(self):
+        """Execute the Ansible Runner Service operations and update the status
+        and result of the underlying Completion object.
+        """
+
+        for op in self.operations:
+            # Execute the right kind of http request
+            try:
+                if op.http_operation == "post":
+                    response = self.ar_client.http_post(op.url, op.payload, op.query_string)
+                elif op.http_operation == "delete":
+                    response = self.ar_client.http_delete(op.url)
+                elif op.http_operation == "get":
+                    response = self.ar_client.http_get(op.url)
+
+                # Any problem executing the secuence of operations will
+                # produce an errored completion object.
+                if response.status_code != requests.codes.ok:
+                    self._status = ExecutionStatusCode.ERROR
+                    self._result = response.text
+                    return self._status
+
+            # Any kind of error communicating with ARS or preventing
+            # to have a right http response
+            except Exception as ex:
```

Comment:
Catching `Exception` is also going to catch exceptions like `ImportError`, `SyntaxError`, `MemoryError`, `RecursionError`, `SystemExit`, `TypeError`.

Reply:
I have done a complete review of error management. It would be nice if you can take a look :-)